src/cogent3/core/new_alignment.py

```diff
@@ -0,0 +1,449 @@
+    return alphabet.to_indices(seq)
+
+
+@seq_index.register
+def _(seq: bytes, alphabet: CharAlphabet) -> numpy.ndarray:
+    return alphabet.to_indices(seq)
+
+
+@seq_index.register
+def _(seq: numpy.ndarray, alphabet: CharAlphabet) -> numpy.ndarray:
+    return seq.astype(alphabet.array_type)
+
+
+@singledispatch
+def process_name_order(
+    correct_names: Union[dict, tuple, list], name_order: tuple
+) -> tuple:
+    """dict (data) for constructor; tuple for SeqData instance"""
+    raise NotImplementedError(
+        f"process_name_order not implemented for type {type(correct_names)}"
+    )
+
+
+@process_name_order.register
+def _(correct_names: dict, name_order: tuple) -> tuple:
+    keys = correct_names.keys()
+    if name_order is None:
+        return tuple(keys)
+    if set(name_order) == set(keys):
+        return name_order
+    raise ValueError("name_order does not match dictionary keys")
+
+
+def _name_order_list_tuple(name_order, correct_names):
+    """List and tuples have the same implementation for dispatch"""
+    if name_order is None:
+        return correct_names
+    if set(name_order) <= set(correct_names):
+        return tuple(name_order)
+    raise ValueError("some names do not match")
+
+
+@process_name_order.register
+def _(correct_names: tuple, name_order: tuple) -> tuple:
+    return _name_order_list_tuple(name_order, correct_names)
+
+
+@process_name_order.register
+def _(correct_names: list, name_order: tuple) -> tuple:
+    return _name_order_list_tuple(name_order, correct_names)
+
+
+class SeqDataView(SliceRecordABC):
+    """
+    A view class for SeqData, providing properties for different
+    representations.
+
+    self.seq is a SeqData() instance, but other properties are a reference to a single
+    seqid only.
+
+    Example
+    -------
+    data = {"seq1": "ACGT", "seq2": "GTTTGCA"}
+    sd = SeqData(data=data)
+    sdv = sd.get_seq_view(seqid="seq1")
+    """
+
+    __slots__ = ("seq", "start", "stop", "step", "_offset", "_seqid", "_seq_len")
+
+    def __init__(
+        self,
+        *,
+        seq: SeqData,
+        start: Optional[int] = None,
+        stop: Optional[int] = None,
+        step: Optional[int] = None,
+        offset: int = 0,
+        seqid: Optional[str] = None,
+        seq_len: Optional[int] = None,
+    ):
+        if step == 0:
+            raise ValueError("step cannot be 0")
+        step = 1 if step is None else step
+
+        self._seq_len = self._checked_seq_len(seq_len)
+        func = _input_vals_pos_step if step > 0 else _input_vals_neg_step
+        start, stop, step = func(self._seq_len, start, stop, step)
+        self.seq = seq
+        self.start = start
+        self.stop = stop
+        self.step = step
+        self._offset = offset
+        self._seqid = seqid
+
+    def _checked_seq_len(self, seq_len) -> int:
+        assert seq_len is not None
+        return seq_len
+
+    @property
+    def _zero_slice(self):
```

Comment:
Shouldn't `self.seq_len` always be the length of the original/parent sequence?

`start=0, end=0` already added so that the `str` representation of a `SeqData._zero_slice matches the `SeqView.seq = ''` equivalent (i.e. all str-representation slices go through `SeqDataView.str_value()`)